### PR TITLE
Fix for Issue-897 [BindBean behaves unexpectedly for property names that start with single letters or acronyms]

### LINKED ISSUE #897: BindBean behaves unexpectedly for property names that start with single letters or acronyms
Per [Introspector.decapitalize() javadoc](http://docs.oracle.com/javase/8/docs/api/java/beans/Introspector.html#decapitalize-java.lang.String-), the property name for a getter named `getPCode()` is `PCode`, not `pCode` as one might expect. So a bound parameter `:pCode` would not be bound to the expected bean property.

If a property name reported by `Introspector` starts with a capital letter, we should _actually_ decapitalize the property name and bind the property if it matches that variant.

### PR DESCRIPTION
**Pull request by hgschmie**
## Summary
- Fixes `@BindBean` behavior for property names where `Introspector.decapitalize()` keeps the first character uppercase (e.g. `getPCode()` yields property name "PCode" instead of "pCode")
- Adds an alternate lowercase-first-char entry in the bean properties map so both `:PCode` and `:pCode` resolve to the same property
- The fix is in `BeanPropertiesFactory.PropertiesHolder` where the properties map is built

Fixes #897

## Test plan
- [x] Test binding with prefixed bean name (`:bean.pCode`)
- [x] Test direct binding (`:pCode`)
- [x] Test original property name still works (`:PCode`)
- [x] Test mapping with `BeanMapper`

Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by hgschmie** [comment]
It does fix a problem we have for a very long time; However we might hold off on merging this for a more comprehensive rewrite of the bean/record/class mapping framework. 

@stevenschlansker ?

File: core/src/main/java/org/jdbi/v3/core/mapper/reflect/internal/BeanPropertiesFactory.java [hunk 1]
```diff
@@ -199,10 +200,27 @@ static class PropertiesHolder<T> {
...
+                        String name = entry.getKey();
+                        if (name.length() > 1 && Character.isUpperCase(name.charAt(0))) {
+                            String altName = Character.toLowerCase(name.charAt(0)) + name.substring(1);
+                            if (!props.containsKey(altName)) {
+                                altNames.put(altName, entry.getValue());
+                            }
+                        }
+                    }
+                    props.putAll(altNames);
+                    properties = props;
```
[the remark this answers is not in the document]
Comment (stevenschlansker) [reply core/src/main/java/org/jdbi/v3/core/mapper/reflect/internal/BeanPropertiesFactory.java:223]: This is a good point that the current approach here does effectively add a new property, rather than simply an alias to an existing property.
I briefly checked the code and did not see an obvious problem, but there could well be one.

**Comment by hgschmie** [comment]
I think that the fix is good, but I still want to hold off. There is no actual ask for this change (the issue is very old and there was never real followup). 

Rewriting the Bean/Field/Constructor mapper framework is basically the last big thing that I want to do with Jdbi, after that I pretty much consider it "done" feature-wise. Getting the config changes in would be great but that is a performance optimization, not a true major code change. 

I want to hold of these changes for that rewrite. I marked it as "on hold" for now unless you feel strongly about applying it.

**Comment by stevenschlansker** [comment]
This is fine, except:

> Getting the config changes in would be great but that is a performance optimization, not a true major code change.

Fixing configuration will definitely require breaking API changes.

**Comment by vibingarthur** [comment]
## Code Review: PR #2925 - Fix for Issue #897

### Summary
This PR fixes GitHub issue #897 by allowing bean properties whose names start with a single uppercase letter followed by another uppercase letter (e.g., getPCode) to be bound using the intuitive lowercase form (:pCode).

### Changes
1. BeanPropertiesFactory.java - Adds alternative property name mapping
2. TestBeanPropertyAcronymBinding.java - Comprehensive tests
3. RELEASE_NOTES.md - Documents the fix

### Review
The code looks good. Key observations:
- The fix correctly handles the edge case where Introspector.decapitalize() does not lowercase the first character when the second is also uppercase
- The solution adds an alternate mapping without removing the original property name, so both :PCode and :pCode work
- Tests are thorough and cover bindbean, direct binding, and mapping scenarios

No changes required - the implementation is clean and well-tested.

---
*Review by pr-reviewer-agent*

**Comment by vibingarthur** [comment]
test

**Comment by vibingarthur** [comment]
# PR #2925 Review: Fix BindBean for acronym property names

**Commit:** `5237af768` — Fixes #897

## Summary

When `Introspector.decapitalize()` processes a getter like `getPCode()`, it returns "PCode" (not "pCode") because both the 1st and 2nd characters are uppercase. The fix adds a lowercase-first-char alias to the properties map so `:pCode` resolves alongside `:PCode`.

## Verdict: Looks good overall

The approach is correct, minimal, and well-tested.

---

### 1. Mutability of `Collectors.toMap()` result (low risk)

**File:** `BeanPropertiesFactory.java:203-222`

`Collectors.toMap()` returns a `HashMap` by implementation, but the Javadoc doesn't guarantee mutability. The code calls `props.putAll(altNames)` on line 222, which requires mutability.

In practice this will always work, but for defensive correctness:



**Severity:** Nitpick

---

### 2. Edge case: all-uppercase property names produce odd aliases

For `getURL()` → Introspector returns "URL" → the fix creates alias "uRL". These aliases are weird but **harmless** — the original "URL" still works.

---

### 3. Test coverage is solid

Four tests cover:
- Prefixed binding (`:bean.pCode`) — ✓
- Direct binding (`:pCode") — ✓
- Original name still works (`:PCode") — ✓
- Mapping from result set — ✓

---

### 4. No parallel fix needed elsewhere

`Introspector.getBeanInfo()` is only used in `BeanPropertiesFactory`. Other `PojoProperties` implementations are unaffected.

---

## Summary

| Item | Severity |
|---|---|
| `Collectors.toMap` mutability | Nitpick (optional fix) |
| All-caps getter aliases | Info only |
| Overall | ✅ Approved |

---

*Review by pr-reviewer-agent*
